pulsar/transaction_coordinator_client.go

```diff
@@ -20,23 +20,330 @@ package pulsar
+			t.connectClosedCh <- &connectionClosed{}
+		}
+	}()
+	t.conn.Store(res.Cnx)
+	t.log.Infof("Transaction handler with transaction coordinator id %d connected", t.partition)
+	return nil
+}
+
+func (t *transactionHandler) getConn() internal.Connection {
+	return t.conn.Load().(internal.Connection)
+}
+
+func (t *transactionHandler) runEventsLoop() {
+	for {
+		select {
+		case <-t.closeCh:
+			return
+		case req := <-t.requestCh:
+			switch r := req.(type) {
+			case *newTxnOp:
+				t.newTransaction(r)
+			case *addPublishPartitionOp:
+				t.addPublishPartitionToTxn(r)
+			case *addSubscriptionOp:
+				t.addSubscriptionToTxn(r)
+			case *endTxnOp:
+				t.endTxn(r)
+			}
+		case <-t.connectClosedCh:
+			t.log.Infof("Transaction handler %d will reconnect to the transaction coordinator", t.partition)
+			t.reconnectToBroker()
+		}
+	}
+}
+
+func (t *transactionHandler) reconnectToBroker() {
+	var delayReconnectTime time.Duration
+	var defaultBackoff = internal.DefaultBackoff{}
+
+	for {
+		if t.getState() == txnHandlerClosed {
+			// The handler is already closing
+			t.log.Info("transaction handler is closed, exit reconnect")
+			return
+		}
+
+		delayReconnectTime = defaultBackoff.Next()
+
+		t.log.WithFields(log.Fields{
+			"delayReconnectTime": delayReconnectTime,
+		}).Info("Reconnecting to broker")
+		time.Sleep(delayReconnectTime)
+
+		// double check
+		if t.getState() == txnHandlerClosed {
+			// Txn handler is already closing
+			t.log.Info("transaction handler is closed, exit reconnect")
+			return
+		}
+
+		err := t.grabConn()
+		if err == nil {
+			// Successfully reconnected
+			t.log.Info("Reconnected transaction handler to broker")
+			return
+		}
+		t.log.WithError(err).Error("Failed to create transaction handler at reconnect")
+		errMsg := err.Error()
+		if strings.Contains(errMsg, errMsgTopicNotFound) {
+			// when topic is deleted, we should give up reconnection.
+			t.log.Warn("Topic Not Found.")
+			break
+		}
+	}
+}
+
+func (t *transactionHandler) checkRetriableError(err error, op any) bool {
+	if err != nil && errors.Is(err, internal.ErrConnectionClosed) {
+		// We are in the EventLoop here, so we need to insert the request back to the requestCh asynchronously.
+		go func() {
+			t.requestCh <- op
+		}()
+		return true
+	}
+	return false
+}
+
+type newTxnOp struct {
+	// Request
+	timeout time.Duration
+
+	// Response
+	done  chan any
+	err   error
+	txnID *TxnID
+}
+
+func (t *transactionHandler) newTransaction(op *newTxnOp) {
+	requestID := t.tc.client.rpcClient.NewRequestID()
+	nextTcID := t.tc.nextTCNumber()
+	cmdNewTxn := &pb.CommandNewTxn{
+		RequestId:     proto.Uint64(requestID),
+		TcId:          proto.Uint64(nextTcID),
+		TxnTtlSeconds: proto.Uint64(uint64(op.timeout.Milliseconds())),
+	}
+	res, err := t.tc.client.rpcClient.RequestOnCnx(t.getConn(), requestID, pb.BaseCommand_NEW_TXN, cmdNewTxn)
+	if t.checkRetriableError(err, op) {
+		return
+	}
+	defer close(op.done)
+	defer t.tc.semaphore.Release()
```

Comment:
If my understand right.

This semaphore will be released 2 times for the same operation.

https://github.com/apache/pulsar-client-go/pull/1237/files#diff-d923c7a298876f490ae5b4b04c712deae3da2afa49cc5161e1c37e70e9822f73R394

Other method have the same issue.

Reply:
Good catch. Yes. I will fix them.